qiita_db/metadata_template/base_metadata_template.py

```diff
@@ -1183,34 +1182,11 @@ def _common_to_dataframe_steps(self, samples=None):
                 sql += ' AND sample_id IN %s'
                 qdb.sql_connection.TRN.add(sql, [tuple(samples)])
 
-            # this query is going to return a tuple
-            # (sample_id, dict of columns/values); however it's important to
-            # notice that we can't assure that all column/values pairs are the
-            # same for all samples as we are not doing full bookkeeping of all
-            # the columns in all the samples. Thus, we have 2 options:
-            # 1. use dict() on the query result with pd.DataFrame.from_dict so
-            #    pandas deals with this; but this takes a crazy amount of time,
-            #    for more info google: "performance pandas from_dict"
-            # 2. generate a matrix rows/samples, cols/values and load them
-            #    via pandas.DataFrame, which actually has good performace
-            data = []
-            for sid, values in qdb.sql_connection.TRN.execute_fetchindex():
-                # creating row of values, first insert sample id
-                vals = [sid]
-                # then loop over all the possible values making sure that if
-                # the column doesn't exist in that sample, it gets a None
-                for c in cols:
-                    v = None
-                    if c in values:
-                        v = values[c]
-                    vals.append(v)
-                # append the row to the full matrix
-                data.append(vals)
-            cols.insert(0, 'sample_id')
-            df = pd.DataFrame(data, columns=cols, dtype=str)
-            df.set_index('sample_id', inplace=True)
-
-            # Make sure that we are changing np.NaN by Nones
+            df = pd.DataFrame(qdb.sql_connection.TRN.execute_fetchindex())
+            df = pd.concat([df.drop(1, axis=1),
+                            pd.DataFrame(df[1].tolist())], axis=1)
```

Comment:
Would it be possible to do `pd.concat([df.drop(1, axis=1), df[1]])`? It not really obvious why the new `DataFrame` construction is needed or the `tolist`. Oh wait, is this supposed to be a reorder operation moving the first column to the end of the frame? If so, then i think this does it:

```python
cols = [df.columns[0]] + df.columns[2:].tolist() + [df.columns[1]]
df = df[cols]
```

Reply:
Good question. The metadata tables are basically 2 columns: sample_id, metadata (json, which is returned as a dict). In other words, the SQL will return [[sample_id, {'col_a': 'a', 'col_b': 'b'}, ...], [sample_id, {'col_a': 'a', 'col_b': 'b'}, ...], ...] so that line basically expands the dict into actual columns in the dataframe, like this:
```python
In [61]: data = ['x', {'a': 1, 'b': 2}], ['y', {'a': 1, 'c': 2}]                                                                                                                                                                                                                            

In [62]: df = pd.DataFrame(data)                                                                                                                                                                                                                                                            

In [63]: df                                                                                                                                                                                                                                                                                 
Out[63]: 
   0                 1
0  x  {'a': 1, 'b': 2}
1  y  {'a': 1, 'c': 2}

In [64]: pd.concat([df.drop(1, axis=1), pd.DataFrame(df[1].tolist())], axis=1)                                                                                                                                                                                                              
Out[64]: 
   0  a    b    c
0  x  1  2.0  NaN
1  y  1  NaN  2.0
```

Reply:
Okay, thanks. What about this? It's a little less juggling and you get the index "for free"? 

```python
>>> import pandas as pd
>>> data = ['x', {'a': 1, 'b': 2}], ['y', {'a': 1, 'c': 2}]
>>> pd.DataFrame([d for _, d in data], index=[i for i, _ in data])
   a    b    c
x  1  2.0  NaN
y  1  NaN  2.0
```

Reply:
Thanks, this made the code a bit faster too. If you agree with the changes, and the tests pass, could you merge?